### PR TITLE
PSR-Tracing - Pre-Draft Specification & Meta

### PR DESCRIPTION
**Pull request by AAllport**
Add Initial documents for Tracing PSR

Discussion: https://groups.google.com/g/php-fig/c/6Nm0ZQ1_hJA

### CODE REVIEW

File: proposed/tracing-meta.md
Comment (PowerKiKi) [inline proposed/tracing-meta.md:15]:
```suggestion
  This would in turn allow other APM providers to receive the same signals for further processing or analysis.
```

File: proposed/tracing-meta.md
Comment (PowerKiKi) [inline proposed/tracing-meta.md:17]:
```suggestion
* This PSR may provide a minimal `TraceProvider`, etc. for other APM tools to extend, should they choose.
```

File: proposed/tracing-meta.md
Comment (PowerKiKi) [inline proposed/tracing-meta.md:23]:
```suggestion
* This PSR does not cover collecting metrics within a codebase, only traces
```

**Comment by AAllport** [comment]
Thanks for those grammar checks @PowerKiKi, all applied!

**Comment by Jean85** [comment]
Question: is it correct to say that this would be a PSR and not a PER? Can we count on OTel not changing in the future?

**Comment by AAllport** [comment]
> Question: is it correct to say that this would be a PSR and not a PER? Can we count on OTel not changing in the future?

Good point. And it had certainly crossed my mind.
Whilst I will ultimately defer to @alekitto, this feels like it would be suitable as a PSR due to the limited scope of the document.

I think it's highly unlikely that OTel will move away from a Tracers->Spans->SubSpan layout.
Given this proposal is aimed at providing a common denominator. Only a drastic change to that methodology would cause this to become _outdated_

I do call on the OTel team to dispute anything I've said there :p
CC: @tidal @brettmc

**Comment by alekitto** [comment]
> Question: is it correct to say that this would be a PSR and not a PER? Can we count on OTel not changing in the future?

Interesting question.

In my opinion yes, this should be a PSR: the goal should be providing a minimal stable set of interfaces the library authors can rely on. Also, OTel spec just reached 1.0, so can be considered stable enough for our goal. Ultimately, declaring that this specification will be _constantly_ changing, will sure cause a negative impact on its adoption.

I will be glad to hear any other opinion on this subject

**Comment by tidal** [comment]
> OTel spec just reached 1.0

To clarify, The Trace "signal" API (which is relevant for us here) is marked as stable in the Otel specification.
The PHP implementation is not yet stable, however we are not planning on any drastic changes regarding 
the API (that's the component were all the relevant interfaces live).
However since all @open-telemetry/php-maintainers are on board with this initiative, we can still make 
adjustments if it makes sense.

> Can we count on OTel not changing in the future?

Yes and No. 
You can have a look at Otel's stability guarantess:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#api-stability
However, we as Otel PHP maintainers have a great interest in this PSR, as it will also help us with adoption.
So we likely don't want to break anything ragarding interfaces.

File: proposed/tracing.md
Comment (Crell) [inline proposed/tracing.md:1]: Could stand to be shorter. :smile:

File: proposed/tracing-meta.md
Comment (AAllport) [reply proposed/tracing-meta.md:28]: My original thought was "Aim" would create that flexibility.
But I agree the statement you have provided is certainly clearer. I will update 🔜 !

**Comment by Crell** [comment]
Note that it's totally fair game for the spec to reference some other existing spec.  Eg, "this value may be any legal mime type as defined by IANA" or "this value may be any tag defined by the OpenTelemetry spec as listed at URL blah".  IMO, that's a fair "loophole" around spec's being immutable, as long as the thing they're linking out to is cleanly extensible and well maintained.  (That would preclude, for instance, using an enum to enforce a fixed list, but if the list is not truly fixed then that wouldn't make sense anyway."

Also, can I just note for the record that I love seeing the OT maintainers working here directly to ensure everything is as consistent and compatible as possible.  That's the ideal case. :grin:

File: proposed/tracing-meta.md
Comment (AAllport) [reply proposed/tracing-meta.md]: Resolved: 3f318e9

File: proposed/tracing.md
Comment (AAllport) [reply proposed/tracing.md]: Resolved: 30401ee

File: proposed/tracing-meta.md
Comment (brettmc) [inline proposed/tracing-meta.md:8]: *libraries

Comment (brettmc) [inline proposed/tracing-meta.md:8]:
```suggestion
This standard aims to create an intentionally minimalist set of interfaces that can be used to provide tracing signals to 3rd party libraries in a unified manner.
```

File: proposed/tracing-meta.md
Comment (brettmc) [inline proposed/tracing-meta.md:28]:
```suggestion
We aim to allow the majority of APM tools to use the PSR interfaces with minimum backwards incompatible changes to encourage adoption.
```
I think this is what is meant...

Comment (AAllport) [reply proposed/tracing-meta.md:28]: 🤦

File: proposed/tracing-meta.md
Comment (brettmc) [inline proposed/tracing-meta.md:18]:
```suggestion
* To provide a set of interfaces for HTTP client implementations to enable sending outgoing distributed trace headers.
```

I think that an important goal of the spec should also be to provide interfaces to support distributed tracing. I see it as separate to providing tracing signals, and I think it's more widely useful. I believe most/all APM implementations support the [W3C traceparent+tracestate headers](https://www.w3.org/TR/trace-context/) (I checked elastic, datadog, and new relic), but they sometimes leave the injection of outgoing trace headers as an exercise for the reader. Having a PSR interface that was implemented by popular http client implementations would make tracing easier and more useful, IMO.

Comment (alekitto) [reply proposed/tracing-meta.md:18]: I'm not sure it fits this proposal: distributed tracing is surely useful and we need to provide interfaces to support it, but the choice of decorating an http request with tracing header should be an implementation detail. 

Additionally this proposal is not an extension of PSR-18 and it shouldn't be from my point of view.

Surely http clients should be aware of the problem, but whether to decorate a request and how to do it is little bit out of scope at the moment.

Comment (brettmc) [reply proposed/tracing-meta.md:18]: That makes sense - perhaps then something lighter like `To expose distributed tracing information to other libraries/tools`? I guess the important part is a consistent way to get those values from a tracing implementation. If we think that the existing wording sufficiently covers this, then I'm happy with the goals as they are.

Comment (alekitto) [reply proposed/tracing-meta.md:18]: Yes, much better to me.
Maybe is already covered by `To allow traces collected by various APM libraries to be reused by other libraries/tools.`, but we can explicit it to make it crystal clear.

Comment (AAllport) [reply proposed/tracing-meta.md:31]: I believe I have added clarification by way of the definitions in 916f0c2
Please do let me know if you disagree, etc

**Review by Jean85 — APPROVED** [review]
Entrance vote passed! https://groups.google.com/g/php-fig/c/GP8ThJsC8ac

**Pull request merged**
